src/hashtable.c

```diff
@@ -0,0 +1,2165 @@
+    int16_t pause_rehash;      /* Non-zero = rehashing is paused */
+    int16_t pause_auto_shrink; /* Non-zero = automatic resizing disallowed. */
+    size_t child_buckets[2];   /* Number of allocated child buckets. */
+    void *metadata[];
+};
+
+typedef struct {
+    hashtable *hashtable;
+    bucket *bucket;
+    long index;
+    uint16_t pos_in_bucket;
+    uint8_t table;
+    uint8_t safe;
+    union {
+        /* Unsafe iterator fingerprint for misuse detection. */
+        uint64_t fingerprint;
+        /* Safe iterator temporary storage for bucket chain compaction. */
+        uint64_t last_seen_size;
+    };
+} iter;
+
+/* The opaque hashtableIterator is defined as a blob of bytes. */
+static_assert(sizeof(hashtableIterator) >= sizeof(iter),
+              "Opaque iterator size");
+
+/* Position, used by some hashtable functions such as two-phase insert and delete. */
+struct position {
+    bucket *bucket;
+    uint16_t pos_in_bucket;
+    uint16_t table_index;
+};
+
+static_assert(sizeof(hashtablePosition) >= sizeof(struct position),
+              "Opaque iterator size");
+
+/* State for incremental find. */
+typedef struct {
+    enum {
+        HASHTABLE_CHECK_ENTRY,
+        HASHTABLE_NEXT_ENTRY,
+        HASHTABLE_NEXT_BUCKET,
+        HASHTABLE_FOUND,
+        HASHTABLE_NOT_FOUND
+    } state;
+    short table;
+    short pos;
+    hashtable *hashtable;
+    bucket *bucket;
+    const void *key;
+    uint64_t hash;
+} incrementalFind;
+
+static_assert(sizeof(hashtableIncrementalFindState) >= sizeof(incrementalFind),
+              "Opaque incremental find state size");
+
+/* Struct used for stats functions. */
+struct hashtableStats {
+    int table_index;                /* 0 or 1 (old or new while rehashing). */
+    unsigned long toplevel_buckets; /* Number of buckets in table. */
+    unsigned long child_buckets;    /* Number of child buckets. */
+    unsigned long size;             /* Capacity of toplevel buckets. */
+    unsigned long used;             /* Number of entries in the table. */
+    unsigned long max_chain_len;    /* Length of longest bucket chain. */
+    unsigned long *clvector;        /* Chain length vector; entry i counts
+                                     * bucket chains of length i. */
+};
+
+/* Struct for sampling entries using scan, used by random key functions. */
+
+typedef struct {
+    unsigned size;  /* Size of the entries array. */
+    unsigned seen;  /* Number of entries seen. */
+    void **entries; /* Array of sampled entries. */
+} scan_samples;
+
+/* --- Memory prefetching --- */
+
+/* Check for GCC version >= 4.9 */
+#if defined(__GNUC__) && (__GNUC__ > 4 || (__GNUC__ == 4 && __GNUC_MINOR__ >= 9))
+#define prefetch(addr) __builtin_prefetch(addr)
+/* Check for Clang version >= 3.6 */
+#elif defined(__clang__) && (__clang_major__ > 3 || (__clang_major__ == 3 && __clang_minor__ >= 6))
+#define prefetch(addr) __builtin_prefetch(addr)
+#else
+#define prefetch(addr) ((void)(addr))
```

Comment:
We already have `valkey_prefetch`, defined in config.h.

Reply:
Can we break config.h into smaller units? Or am I over-thinking?